core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java

```diff
@@ -154,23 +163,28 @@ public LoggingService(Service<I, O> delegate, LogLevel level) {
     @Override
     public O serve(ServiceRequestContext ctx, I req) throws Exception {
         if (sampler.isSampled(ctx)) {
-            ctx.log().addListener(log -> logRequest(((ServiceRequestContext) log.context()).logger(),
+            ctx.log().addListener(log -> logRequest(getActiveLogger((ServiceRequestContext) log.context()),
                                                     log, requestLogLevel, requestHeadersSanitizer,
                                                     requestContentSanitizer, requestTrailersSanitizer),
                                   RequestLogAvailability.REQUEST_END);
-            ctx.log().addListener(log -> logResponse(((ServiceRequestContext) log.context()).logger(), log,
-                                                     requestLogLevel,
-                                                     requestHeadersSanitizer,
-                                                     requestContentSanitizer,
-                                                     requestTrailersSanitizer,
-                                                     successfulResponseLogLevel,
-                                                     failedResponseLogLevel,
-                                                     responseHeadersSanitizer,
-                                                     responseContentSanitizer,
-                                                     responseTrailersSanitizer,
-                                                     responseCauseSanitizer),
-                                  RequestLogAvailability.COMPLETE);
+            ctx.log().addListener(
+                    log -> logResponse(getActiveLogger((ServiceRequestContext) log.context()), log,
+                                       requestLogLevel,
+                                       requestHeadersSanitizer,
+                                       requestContentSanitizer,
+                                       requestTrailersSanitizer,
+                                       successfulResponseLogLevel,
+                                       failedResponseLogLevel,
+                                       responseHeadersSanitizer,
+                                       responseContentSanitizer,
+                                       responseTrailersSanitizer,
+                                       responseCauseSanitizer),
+                    RequestLogAvailability.COMPLETE);
         }
         return delegate().serve(ctx, req);
     }
+
+    private Logger getActiveLogger(ServiceRequestContext context) {
+        return logger != null ? logger : context.logger();
+    }
```

Comment:
How about removing this method and using [`firstNonNull(logger, context.logger())`](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Objects.html#firstNonNull(T,%20T))?

Reply:
fixed